flasc/analysis/expected_power_analysis.py

```diff
@@ -0,0 +1,789 @@
+        df_freq_pl (pl.DataFrame): A polars dataframe with the frequency of each bin.
+            Defaults to None.
+        remove_any_null_turbine_bins (bool): When computing farm power, remove any bins where
+            and of the test turbines is null.  Defaults to False.
+        N (int): The number of bootstrap samples. Defaults to 1.
+        percentiles (List[float]): The percentiles to calculate for the bootstrap samples.
+            Defaults to [2.5, 97.5].
+
+    Returns:
+        Tuple[pl.DataFrame, pl.DataFrame, Dict[str, Dict[str, float]]]: A tuple containing the
+             binned dataframe, the summed dataframe, and the uplift results.
+    """
+    uplift_single_outs = [
+        _total_uplift_expected_power_single(
+            a_in.resample_energy_table(perform_resample=(i != 0)),
+            test_cols=test_cols,
+            wd_cols=wd_cols,
+            ws_cols=ws_cols,
+            uplift_pairs=uplift_pairs,
+            uplift_names=uplift_names,
+            wd_step=wd_step,
+            wd_min=wd_min,
+            wd_max=wd_max,
+            ws_step=ws_step,
+            ws_min=ws_min,
+            ws_max=ws_max,
+            bin_cols_in=bin_cols_in,
+            weight_by=weight_by,
+            df_freq_pl=df_freq_pl,
+            remove_any_null_turbine_bins=remove_any_null_turbine_bins,
+        )
+        for i in range(N)
+    ]
+
+    # Add in the statistics
+    bootstrap_uplift_result = {}
+
+    for uplift_name in uplift_names:
+        uplift_values = np.zeros(N)
+
+        for i in range(N):
+            uplift_values[i] = uplift_single_outs[i][2][uplift_name]
+
+        uplift_central = uplift_values[0]
+        uplift_lb = np.quantile(uplift_values, percentiles[0] / 100)
+        uplift_ub = np.quantile(uplift_values, percentiles[1] / 100)
+
+        bootstrap_uplift_result[uplift_name] = {
+            "energy_uplift_ctr": uplift_central,
+            "energy_uplift_lb": uplift_lb,
+            "energy_uplift_ub": uplift_ub,
+            "energy_uplift_ctr_pc": (uplift_central - 1) * 100,
+            "energy_uplift_lb_pc": (uplift_lb - 1) * 100,
+            "energy_uplift_ub_pc": (uplift_ub - 1) * 100,
+        }
+
+    return uplift_single_outs[0][0], uplift_single_outs[0][1], bootstrap_uplift_result
+
+
+def _total_uplift_expected_power_with_standard_error(
+    df_: pl.DataFrame,
+    test_cols: List[str],
+    wd_cols: List[str],
+    ws_cols: List[str],
+    uplift_pairs: List[Tuple[str, str]],
+    uplift_names: List[str],
+    wd_step: float = 2.0,
+    wd_min: float = 0.0,
+    wd_max: float = 360.0,
+    ws_step: float = 1.0,
+    ws_min: float = 0.0,
+    ws_max: float = 50.0,
+    bin_cols_in: List[str] = ["wd_bin", "ws_bin"],
+    weight_by: str = "min",  # min or sum
+    df_freq_pl: pl.DataFrame = None,
+    percentiles: List[float] = [2.5, 97.5],
+    remove_any_null_turbine_bins: bool = False,
+    set_cov_to_zero_or_var: str = "zero",
+    use_cov_when_available: bool = False,
```

Comment:
Since we're only allowing three options for filling in covariance terms, I say we use one argument instead of the two above. It could be cov_terms, with options "zero", "var", and "cov". In the docstring we can describe how the "cov" option will fill in undefined terms with the product of variances.

Reply:
Nice suggestion!  I made teh change